bundles/alpha.model/src/alpha/model/util/AffineFactorizer.xtend

```diff
@@ -0,0 +1,391 @@
+			updatedMatrix = updatedMatrix.setElement(i, col, coefficient)
+		}
+		
+		// Copy the coefficients for the input variables.
+		val inCount = expression.dim(ISLDimType.isl_dim_in)
+		for (i : 0 ..< inCount) {
+			val coefficient = expression.getCoefficientVal(ISLDimType.isl_dim_in, i)
+			updatedMatrix = updatedMatrix.setElement(paramCount + i, col, coefficient)
+		}
+		
+		// Finally, copy the constant.
+		val constant = expression.constantVal
+		updatedMatrix = updatedMatrix.setElement(paramCount + inCount, col, constant)
+		
+		return updatedMatrix
+	}
+	
+	
+	////////////////////////////////////////////////////////////
+	// Hermite Decomposition
+	////////////////////////////////////////////////////////////
+	
+	/**
+	 * Performs the column-oriented Hermite decomposition of the given
+	 * matrix, returning H and Q as the two outputs (as a key->value pair,
+	 * in that order respectively). Any columns of 0's have been dropped
+	 * from the right of H, along with the same number of rows from the
+	 * bottom of Q.
+	 */
+	def static hermiteDecomposition(ISLMatrix matrix) {
+		val hermiteResult = matrix.leftHermite
+		return reduceHermiteDimensionality(hermiteResult.h, hermiteResult.q)
+	}
+
+	/**
+	 * Reduces the dimensionality of the matrices constructed by the
+	 * Hermite Normal Form calculation.
+	 * This is done by dropping the columns of zeros from the right of H
+	 * along with the same number of rows from the bottom of Q.
+	 * This assumes that all columns of zeros are on the right of H,
+	 * which matches ISL's implementation.
+	 */
+	def static reduceHermiteDimensionality(ISLMatrix h, ISLMatrix q) {
+		val emptyCols = countEmptyCols(h)
+		if (emptyCols == 0) {
+			return h -> q
+		}
+
+		val firstToDrop = h.nbCols - emptyCols
+		val hUpdated = h.dropCols(firstToDrop, emptyCols)
+		val qUpdated = q.dropRows(firstToDrop, emptyCols)
+
+		return hUpdated -> qUpdated
+	}
+
+	/**
+	 * Returns the number of empty columns in the given matrix.
+	 * Assumes that the empty rows appear only on the right of the matrix.
+	 */
+	def private static countEmptyCols(ISLMatrix matrix) {
+		// Note: Don't check if column 0 is empty, as we never want a situation
+		// where we've dropped every single column, as ISL doesn't like that.
+		val cols = matrix.nbCols
+		val emptyRowCount = (cols >.. 1).takeWhile[col | isColEmpty(matrix, col)].size
+		return emptyRowCount
+	}
+
+	/** Checks if a column of the given matrix is empty or not. */
+	def private static isColEmpty(ISLMatrix matrix, int col) {
+		val rows = matrix.nbRows
+		return ! (0 ..< rows).exists[row | matrix.getElement(row, col) != 0]
+	}
+	
+	
+	////////////////////////////////////////////////////////////
+	// Matrix to Expression Conversion
+	////////////////////////////////////////////////////////////
+	
+	/** Converts a matrix into an expression. Each column is for one of the output dimensions. */
+	def static matrixToExpression(ISLMatrix matrix, ISLSpace space) {
```

Comment:
This (and the method below `columnToExpression`) just takes an ISLMatrix as input and produces its corresponding ISLMultiAff, right?

There is already a function to go from an alpha.model.matrix.Matrix -> ISLMultiAff. Can we just use this? We’d have to convert the ISLMatrix to an alpha.model.matrix.Matrix object first of course but I think it better to have two function calls (the first from ISLMatrix -> alpha.model.matrix.Matrix, and a second from alpha.model.matrix.Matrix -> ISLMultiAff) rather than ~50 lines that touch the ISLAff internal coefficient values themselves. +maintability +correctness (there ought to be fewer edges cases)

I just hate to put in more code that does the same thing, just slightly differently.